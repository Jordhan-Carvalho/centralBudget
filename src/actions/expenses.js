import db from '../firebase/firebase';


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
  });

    
  export const startAddExpense = (expenseData = {}) => {
    // only works bc redux middleware (redux thunk)
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0,
      } = expenseData;
      const expense = {description, note, amount, createdAt};
      //save the data to firebase (aynsc call) (used the return to testing porpouse)
        const ref = await db.ref(`users/${uid}/expenses`).push(expense);
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    };
  };
  
  // REMOVE_EXPENSE
  export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
  });

  export const startRemoveExpense = (expense) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
        const id = expense.id;
  try{
  await db.ref(`users/${uid}/expenses/${id}`).remove();
        dispatch(removeExpense(expense));
      } catch (error) {
        alert(error);
      }
      };
      };
  
  // EDIT_EXPENSE
  export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
  });

  export const startEditExpense = (id, updates) => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
      try{
        await db.ref(`users/${uid}/expenses/${id}`).update(updates);
        dispatch(editExpense(id, updates));
      } catch (error) {
        console.log(error);
      }
    };
  };

  // SET_EXPENSES
  export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
  });

  // export const startSetExpenses
  export const startSetExpenses = () => {
    return async (dispatch, getState) => {
      const uid = getState().auth.uid;
  // 1. fetch all expense data once
  //2. Parse the data to array
  const expenses = []; 
  await db.ref(`users/${uid}/expenses`).once('value', (snapshot) => {
        snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
        });
        // 3. dispatch SET_EXPENSES
        dispatch(setExpenses(expenses));
      };
      };