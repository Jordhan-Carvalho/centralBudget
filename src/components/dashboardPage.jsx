import React, { Component } from "react";
import TransChart from "./transChart";
import SavingsChart from "./savingsChart";
import ListFilters from "./common/listFilters";
import TransSumary from "./common/transSumary";

class DashboardPage extends Component {
  render() {
    return (
      <React.Fragment>
        <div className="row justify-content-center">
          <div className="col-12 boxdash">
            <TransChart height={200} />
          </div>
        </div>

        <div className="row justify-content-center">
          <div className="col-12 boxdash">
            <SavingsChart height={200} />
          </div>
        </div>

        <div className="row">
          <div className="col-12 boxdash">
            <ListFilters />
            <TransSumary type={"expense"} />
            <TransSumary type={"income"} />
            <TransSumary type={"saving"} />
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export default DashboardPage;
