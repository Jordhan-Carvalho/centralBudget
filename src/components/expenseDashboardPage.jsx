import React, { Component } from "react";
import { Link } from "react-router-dom";
import ExpenseList from "./expenseList";
import ExpenseListFilters from "./expenseListFilters";

class ExpenseDashboardPage extends Component {
  state = {};
  render() {
    return (
      <React.Fragment>
        <h1>Hello</h1>
        <ExpenseListFilters />
        <ExpenseList />
        <Link to="/create">Add expense</Link>
        <Link to="edit">edit expense</Link>
      </React.Fragment>
    );
  }
}

export default ExpenseDashboardPage;
