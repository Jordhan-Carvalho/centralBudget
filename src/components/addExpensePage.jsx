import React, { Component } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./expenseForm";
import { addExpense } from "../actions/expenses";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add exprense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    addExpense: expense => dispatch(addExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
