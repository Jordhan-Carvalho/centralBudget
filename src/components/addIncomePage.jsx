import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "./common/form";
import { startAddIncome } from "../actions/incomes";

export class AddIncomePage extends Component {
  onSubmit = income => {
    this.props.startAddIncome(income);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Income</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddIncome: income => dispatch(startAddIncome(income))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddIncomePage);
