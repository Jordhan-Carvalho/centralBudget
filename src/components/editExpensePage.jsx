import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "./common/form";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends Component {
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense);
    this.props.history.push("/");
  };

  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div className="form-group">
        <h1>Edit {this.props.expense.description}</h1>
        <div className="form-group">
          <Form expense={this.props.expense} onSubmit={this.onSubmit} />
        </div>
        <div className="form-group">
          <button className="btn btn-danger" onClick={this.onRemove}>
            Remove
          </button>
        </div>
      </div>
    );
  }
}

// maps

const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (expenseID, expense) =>
      dispatch(startEditExpense(expenseID, expense)),
    startRemoveExpense: expense => dispatch(startRemoveExpense(expense))
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id;
    })
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
