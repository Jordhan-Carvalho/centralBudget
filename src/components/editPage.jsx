import React, { Component } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./expenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditPage extends Component {
  onRemove = () => {
    this.props.removeExpense(this.props.expense);
    this.props.history.push("/");
  };

  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button className="btn btn-primary" onClick={this.onRemove}>
          Remove
        </button>
      </div>
    );
  }
}

// maps

const mapDispatchToProps = dispatch => {
  return {
    editExpense: (expenseID, expense) =>
      dispatch(editExpense(expenseID, expense)),
    removeExpense: expense => dispatch(removeExpense(expense))
  };
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => {
      return expense.id === props.match.params.id;
    })
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPage);
