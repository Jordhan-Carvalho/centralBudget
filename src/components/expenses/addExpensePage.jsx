import React, { Component } from "react";
import { connect } from "react-redux";
import Form from "../common/form";
import { startAddExpense } from "../../actions/expenses";

export class AddExpensePage extends Component {
  onSubmit = expense => {
    this.props.startAddExpense(expense);
    this.props.history.push("/expenses");
  };

  render() {
    return (
      <React.Fragment>
        <h1>Add Expense</h1>
        <Form onSubmit={this.onSubmit} />
      </React.Fragment>
    );
  }
}

// to simplify props.dispatch(addExpense(expense)) to props.onSubmit(expense)
const mapDispatchToProps = dispatch => {
  return {
    startAddExpense: expense => dispatch(startAddExpense(expense))
  };
};

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
