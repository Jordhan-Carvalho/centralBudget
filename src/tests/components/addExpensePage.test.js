import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/addExpensePage';
import expenses from '../fixtures/expenses';

// using lifecycle methods (global) refactoring
let addExpense, history, wrapper;
beforeEach(() => {
    addExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />);
});
//

test('should render AddExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

//test in case of submiting everything right
test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    //check if spies (mock fn) are ij
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenLastCalledWith(expenses[0]);
});