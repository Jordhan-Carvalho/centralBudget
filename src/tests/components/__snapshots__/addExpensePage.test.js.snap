// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddExpensePage correctly 1`] = `
<Fragment>
  <h1>
    Add exprense
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</Fragment>
`;
